src/qobj/arithmetic_and_attributes.jl

```diff
@@ -586,23 +586,43 @@ purity(ρ::QuantumObject{<:AbstractArray{T},OperatorQuantumObject}) where {T} =
 @doc raw"""
     tidyup(A::QuantumObject, tol::Real=1e-14)
 
-Removes those elements of a QuantumObject `A` whose absolute value is less than `tol`.
+Given a [`QuantumObject`](@ref) `A`, check the real and imaginary parts of each element separately. Remove the real or imaginary value if its absolute value is less than `tol`.
 """
 tidyup(A::QuantumObject{<:AbstractArray{T}}, tol::T2 = 1e-14) where {T,T2<:Real} =
     QuantumObject(tidyup(A.data, tol), A.type, A.dims)
-tidyup(A::AbstractArray{T}, tol::T2 = 1e-14) where {T,T2<:Real} = @. T(abs(A) > tol) * A
-tidyup(A::AbstractSparseMatrix{T}, tol::T2 = 1e-14) where {T,T2<:Real} = droptol!(copy(A), tol)
+tidyup(A::AbstractArray{T}, tol::T2 = 1e-14) where {T,T2<:Real} = _DropTol(A, tol)
+function tidyup(A::AbstractSparseMatrix{T}, tol::T2 = 1e-14) where {T,T2<:Real}
+    A2 = copy(A)
+    _DropTol!(A2.nzval, tol)
```

Comment:
calling `A2.nzval` won’t work for CUDA arrays, where they used a different name. I remember that there is a specific function (`nonzeros(A2)`?) for that, which works using multiple dispatch.

Reply:
Done~